machinery/topology_test.go

```diff
@@ -369,3 +397,72 @@ func TestTopologyWithRuntimeObjects(t *testing.T) {
 
 	SaveToOutputDir(t, topology.ToDot(), "../tests/out", ".dot")
 }
+func TestTopologyHasLoops(t *testing.T) {
+	apples := []*Apple{{Name: "apple-1"}, {Name: "apple-3"}}
+	oranges := []*Orange{
+		{Name: "orange-1", AppleParents: []string{"apple-1"}},
+		{Name: "orange-3", AppleParents: []string{"apple-3", "apple-1"}},
+		{Name: "orange-4", AppleParents: []string{"apple-3"}},
+	}
+	peaches := []*Peach{
+		{Name: "peach-2", OrangeParents: []string{"orange-1"}},
+		{Name: "peach-1", OrangeParents: []string{"orange-1"}, ChildApples: []string{"apple-1"}},
+		{Name: "peach-2", OrangeParents: []string{"orange-3", "orange-4"}},
+	}
+
+	lemons := []*Lemon{
+		{Name: "lemon-1", PeachParents: []string{"peach-1"}},
+	}
+	_, err := NewTopology(
+		WithTargetables(lemons...),
+		WithTargetables(apples...),
+		WithTargetables(peaches...),
+		WithTargetables(oranges...),
+		WithLinks(
+			LinkPeachesToLemons(peaches),
+			LinkApplesToOranges(apples),
+			LinkPeachesToApples(peaches),
+			LinkOrangesToPeaches(oranges),
+		),
+	)
+	if err == nil {
+		t.Errorf("Expected error, got none")
+	}
+	if err != nil && !strings.Contains(err.Error(), "loop detected") {
+		t.Errorf("Expected loop detection error, got: %s", err.Error())
+	}
```

Comment:
Do we want to save the topology to file in the new tests like in the other tests? 🤔 

```suggestion
	topology, err := NewTopology(
		WithTargetables(lemons...),
		WithTargetables(apples...),
		WithTargetables(peaches...),
		WithTargetables(oranges...),
		WithLinks(
			LinkPeachesToLemons(peaches),
			LinkApplesToOranges(apples),
			LinkPeachesToApples(peaches),
			LinkOrangesToPeaches(oranges),
		),
	)
	if err == nil {
		t.Errorf("Expected error, got none")
	}
	if err != nil && !strings.Contains(err.Error(), "loop detected") {
		t.Errorf("Expected loop detection error, got: %s", err.Error())
	}
	
	SaveToOutputDir(t, topology.ToDot(), "../tests/out", ".dot")
```

Reply:
I am not sure if I want to be saving the image every time the test runs. If it passes why save the graph. Unit test should be fast and file saves seem like a bad idea. Now that said maybe adding the graph saving into the error states of the test would be a good idea.

Reply:
Fair, I've no strong opinion on this. Just thought I'll mention it since some other tests do this

Reply:
Other tests are saving the topology to file. I personally think it would be nice to see how the loops are represented in the graph.